include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,527 @@
+            {
+                philox_kernel();
+                increase_counter_internal();
+                state_.idx = 0;
+            }
+            __loc_result[__elm_count] = state_.Y[state_.idx];
+        }
+
+        return __loc_result;
+    }
+
+    /* generate_internal() specified for sycl_vec output */
+    template <unsigned int _N = 0>
+    std::enable_if_t<(_N > 0), result_type>
+    generate_internal()
+    {
+        result_type __loc_result;
+        for (int __elm_count = 0; __elm_count < _N; __elm_count++)
+        {
+            state_.idx++;
+
+            // check if buffer is empty
+            if (state_.idx == word_count)
+            {
+                philox_kernel();
+                increase_counter_internal();
+                state_.idx = 0;
+            }
+            __loc_result[__elm_count] = state_.Y[state_.idx];
+        }
+
+        return __loc_result;
+    }
+
+    /* generate_internal() specified for a scalar output */
+    template <unsigned int _N = 0>
+    std::enable_if_t<(_N == 0), result_type>
+    generate_internal()
+    {
+        ++state_.idx;
+        if (state_.idx == word_count)
+        { 
+            philox_kernel();
+            increase_counter_internal();
+            state_.idx = 0;
+        }
+
+        return state_.Y[state_.idx];
+    }
+
+    void
+    discard_internal(unsigned long long __z)
+    {
+        std::uint32_t available_in_buffer = word_count - 1 - state_.idx;
+        if (__z <= available_in_buffer) {
+            state_.idx += __z;
+        }
+        else {
+            __z -= available_in_buffer;
+            int tail = __z % word_count;
+            if (tail == 0) {
+                increase_counter_internal(__z / word_count);
+                state_.idx = word_count - 1;
+            }
+            else
+            {
+                if (__z > word_count) {
+                    increase_counter_internal((__z - 1) / word_count);
+                }
+                philox_kernel();
+                increase_counter_internal();
+                state_.idx = tail - 1;
+            }
+        }
+    }
+
+    /* Internal generation Philox kernel */
+    void
+    philox_kernel()
+    {
+        if constexpr (word_count == 2)
+        {
+            scalar_type __V0 = (state_.X[0]) & in_mask;
+            scalar_type __V1 = (state_.X[1]) & in_mask;
+            scalar_type __K0 = (state_.K[0]);
+            for (std::size_t __i = 0; __i < round_count; ++__i)
+            {
+                auto [__hi0, __lo0] = mulhilo(__V0, multipliers[0]);
+                __V0 = __hi0 ^ __K0 ^ __V1;
+                __V1 = __lo0;
+                __K0 = (__K0 + round_consts[0]) & in_mask;
+            }
+            state_.Y[0] = __V0;
+            state_.Y[1] = __V1;
+        }
+        else if constexpr (word_count == 4)
+        {
+            // permute X to V
+            scalar_type __V2 = (state_.X[0]) & in_mask; // ToDo to check if we need mask
```

Comment:
ToDo left here, I assume we don't need to do & in_mask as we do it in every init/increment_counter/set_counter

Reply:
>  I assume we don't need to do & in_mask as we do it in every init/increment_counter/set_counter

Agree. Also unnecessary comment was removed.